### ISSUE TITLE
Neon.Service: Verify service dependencies

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#1042]
The Istio sidecar can take some time to start which results in the app container being unable to establish outbound network connections during this time and it's not uncommon for apps to log a failure and/or exit when this happens.

It would be nice if the NeonService class would build in the ability to wait until an external connection can be established (perhaps to the Kubernetes API service) before letting the application continue to the running state. This would only happen only when the container is running on Kubernetes, which we'd need to detect somehow.

It looks like Kubernetes injects some environment variables when it starts pod containers:

KUBERNETES_SERVICE_HOST=10.0.0.1
KUBERNETES_SERVICE_PORT=443
KUBERNETES_SERVICE_PORT_HTTPS=443

https://kubernetes.io/docs/concepts/services-networking/connect-applications-service/#environment-variables

This is exactly what I'm looking for, I'll just look for the existence of KUBERNETES_SERVICE_HOST and I'll probably expose this as NeonHelper property.

**Comment by jefflill** [comment]
I discussed this with @marcusbooyah this morning and we realized that this problem is more general than just waiting for the Isto sidecar to be able to proxy network traffic.  It's also very possible that external services required by a service may not be ready at the time the your service is started.  This is common in Kubernetes where a collection of related services are started at the same time and initialize themselves in dependently with different timing.

For example, it's often the case that a database container may take much longer to initialize than a stateless website container will take to start.  The impact is that the website will start almost immediately and see connection failures for sometime (it's not uncommon for database clusters to take 30 or more seconds to discover the other nodes, elect a leader, and generally initialize themselves).  During this time, the website will probably exit with an error with Kubernetes restarting the website until the database connection can finally be established.  In theory, the website container could be smarter about retrying before exiting, but this does not seem to be a common pattern.

A better approach (for `NeonService` at least) is to optionally allow the developer to specify a set of services dependencies by network address that need to be running before the service will be allowed to start.  `NeonService` would attempt to establish socket connections to these addresses within `RunAsync()` but before calling the service's `OnRunAsync()` method to start the service.  We'd also have a generous configurable timeout that will terminate the service when exceeded.

I'm going add a `NeonService.Dependencies` property initialized to an instance of:
```
public class ServiceDependencies
{
    public List<string> Uris = new List<string>();
    public TimeSpan Timeout = TimeSpan.FromSeconds(120);
}
``` 
The developer would initialize the `Uris` list with the network endpoints for the required services.  We'd support the following URI schemes:

- http://host[:port]
- https://host[:port]
- tcp://host:port

We're The PATH part of the HTTP/S URIs would be ignored; we're simply going to attempt socket connections to the host/port and consider the service ready if the connection succeeds.  We're not going to actually submit HTTP/S requests and examine status codes (perhaps for the future if this ends up being important).

We'll also support configuring the `ServiceDependencies` properties from environment variables:
```
NEON_SERVICE_DEPENDENCIES_URIS=http://foo.com;tcp://10.0.0.55:1234
NEON_SERVICE_DEPENDENCIES_TIMEOUT_SECONDS=30
```
The URIs here will be separated by semicolons (;) and any URIs will be **added* to any URIs explicitly specified by the developer in code.

Any environment variable timeout value will **override** anything the developer specified.

The cool thing about the environment variables is that they can be specified in the Helm chart when the pod/service is deployed or even in the container's Dockerfile.  So the combination of this plus being able to specify dependencies in code should be quite flexible.
